### PR TITLE
Return empty CoordinateSequence instead of list

### LINKED ISSUE #959: LineString().coords should return an empty CoordinateSequence, not list
Reported at https://github.com/Toblerity/Shapely/pull/955#discussion_r467715843.

### PR DESCRIPTION
**Pull request by chadwhawkins**
Resolves #959

### CODE REVIEW

**Comment by sgillies** [comment]
@chadwhawkins really nice work! I'm tagging @jorisvandenbossche for review too since this would land in shapely 2.0.

**Comment by jorisvandenbossche** [comment]
@chadwhawkins thanks for working on this!

@sgillies it's not something we also want to change for Shapely 1.8? (you added the 1.8 milestone on the issue)
I am a bit ambivalent on doing it already in 1.8 versus keeping the change for 2.0 (for me it is fine for 1.8 as well). But if it's for 2.0, we need to change the target branch of the PR.

File: tests/test_linestring.py [hunk 1]
```diff
@@ -69,10 +70,12 @@ def test_from_generator():
 def test_from_empty():
     line = LineString()
     assert line.is_empty
+    assert isinstance(line.coords, CoordinateSequence)
     assert line.coords[:] == []
 
     line = LineString([])
     assert line.is_empty
+    assert isinstance(line.coords, CoordinateSequence)
     assert line.coords[:] == []
```
Comment (jorisvandenbossche) [inline tests/test_linestring.py:79]: Could you also add a test for doing `np.array(line.coords)` for an empty LineString? 
You can add it to one of the tests below that already require numpy (and the same for Point and LinearRing)

Comment (chadwhawkins) [reply tests/test_linestring.py:79]: No problem, done. Also corrected an issue with a test class name and a test constructing a LineString when it should have been a LinearRing.

File: tests/test_linestring.py [hunk 2]
```diff
@@ -216,6 +216,12 @@ def test_numpy_coords(self):
         la = np.asarray(line.coords)
         assert_array_equal(la, expected)
 
+        # Check empty
+        line = LineString([])
+        la = np.asarray(line.coords)
+
+        self.assertEqual(la.shape[0], 0)
```
Comment (sgillies) [inline tests/test_linestring.py:223]: @chadwhawkins would you be willing to make this assertion into a separate test function? Like a `test_empty_linestring_coords`? Test isolation is super important to us. We don't want a failure at line 217 to block executing the not related assertion at line 223.

**Review by sgillies — CHANGES_REQUESTED** [review]
@chadwhawkins thank you very much! I have one specific change request and there are a few other places where it should be applied. cc @jorisvandenbossche

File: tests/test_linestring.py (hunk 2 above)
Comment (sgillies) [reply tests/test_linestring.py:223]: I admit that shapely is not the best example historically of isolated test cases, but we're aspiring to do better :smile:

**Comment by jorisvandenbossche** [comment]
@sgillies did you see my question above(https://github.com/Toblerity/Shapely/pull/995#issuecomment-710010645) about targetting 1.8 vs 2.0 for this PR?

**Comment by sgillies** [comment]
@jorisvandenbossche I missed that. Yes, I agree :+1:

**Comment by chadwhawkins** [comment]
@sgillies I went through and put those new assertions into their own tests. Please let me know if that matches what you had in mind.

File: tests/test_linestring.py [hunk 3]
```diff
@@ -216,6 +216,10 @@ def test_numpy_coords(self):
         la = np.asarray(line.coords)
         assert_array_equal(la, expected)
 
+    @unittest.skipIf(not numpy, 'Numpy required')
+    def test_numpy_empty_linestring_coords(self):
+        import numpy as np
```
Comment (sgillies) [inline tests/test_linestring.py:221]: This line can be deleted if we use `numpy` instead of `np` on line 225.

File: tests/test_polygon.py [hunk 4]
```diff
@@ -104,6 +104,10 @@ def test_linearring_coords_numpy():
     expected = np.asarray([(0.0, 0.0), (0.0, 1.0), (1.0, 1.0), (0.0, 0.0)])
     assert_array_equal(ra, expected)
 
+
+def test_numpy_empty_linearring_coords():
+    np = pytest.importorskip("numpy")
```
Comment (sgillies) [inline tests/test_polygon.py:109]: Yes, this is fine, too. It seems like standardizing on one form of the other of skipping without numpy would have benefits, but I don't want to belabor this PR. I'm okay with a mix of `importorskip` and `skipif(not numpy)`. @jorisvandenbossche do you have a strong preference?

Comment (chadwhawkins) [reply tests/test_polygon.py:109]: @sgillies agreed that it's certainly not ideal, basically I just tried to follow the usage elsewhere in the file. Perhaps in a separate PR I could work on transitioning all of the tests to pytest style?

Comment (jorisvandenbossche) [reply tests/test_polygon.py:109]: I have a preference to use pytest whenever we write new tests. 

Doing a large conversion to pytest style might also be welcome, but might maybe be most useful if it directly refactors the tests (like splitting up more, better coverage, .. like eg what I did in https://github.com/Toblerity/Shapely/pull/963). There are also a lot of `self.assertEqual(..)`s to clean up ..

File: tests/test_linestring.py [hunk 5]
```diff
@@ -202,17 +205,24 @@ def test_numpy(self):
...
+        la = numpy.asarray(line.coords)
         assert_array_equal(la, expected)
 
+    @unittest.skipIf(not numpy, 'Numpy required')
+    def test_numpy_empty_linestring_coords(self):
+        # Check empty
+        line = LineString([])
+        la = numpy.asarray(line.coords)
+
+        self.assertEqual(la.shape[0], 0)
```
Comment (jorisvandenbossche) [inline tests/test_linestring.py:224]: Can you assert here the full shape instead of only the first element? Is it actually (0,) or (0, 2) ?

Comment (jorisvandenbossche) [inline tests/test_linestring.py:224]:
```suggestion
        assert la.shape[0] == 0
```

(for *new* tests, I have been using pytest style assertions lately)

File: tests/test_polygon.py (hunk 4 above)
Comment (jorisvandenbossche) [reply tests/test_polygon.py:109]: BTW, specifically for numpy, it's also going to be a requied dependency for 2.0, so then we won't need those skips anymore.

**Comment by chadwhawkins** [comment]
@jorisvandenbossche @sgillies I think I've addressed your comments, mind having another look?

**Comment by jorisvandenbossche** [comment]
Thanks!

It can maybe be left for a follow-up, but should we ensure the shape of the coordinates array for empty geometries has a consistent dimension as for non-empty geometries (so `(0, 2)` instead of `(0,)`) ?

**Review by sgillies — APPROVED** [review]
Thanks!

**Pull request merged**